### PR TITLE
Restore -Wunused to CMake build and add dev mode to disable

### PR DESCRIPTION
**Pull request by GMNGeoffrey**
Avoiding unused variables is good for committed code and really annoying
as a build error when doing interactive builds. We let users distinguish
between these two situations with the `IREE_DEV_MODE` option. The CI
will enforce the strictest options. We may later add more things that
are controlled by this option, so it's deliberately broad to encompass
anything under "setting defaults to be better for development".

Includes fixing `iree_select_compiler_opts` to actually append to the
variable it's given (as the documentation says) rather than overwriting
it.

### CODE REVIEW

**Comment by GMNGeoffrey** [comment]
Depends on https://github.com/google/iree/pull/7003 to clean up existing unused variables in EmitC

File: CMakeLists.txt [hunk 1]
```diff
@@ -193,6 +193,7 @@ if(${IREE_ENABLE_CCACHE})
   endif()
 endif()
 
+option(IREE_DEV_MODE "Configure settings to optimize for IREE development (as opposed to CI or release)" OFF)
```
Comment (marbre) [inline CMakeLists.txt:196]: You may want to document this in https://github.com/google/iree/blob/main/docs/developers/get_started/cmake_options_and_variables.md?

Comment (GMNGeoffrey) [reply CMakeLists.txt:196]: I mean I don't *want* to, but I definitely should :stuck_out_tongue:

Comment (marbre) [reply CMakeLists.txt:196]: As my reviews are optional anyway/don't approve anything, I tried to phrase it politely :stuck_out_tongue_winking_eye:

Comment (GMNGeoffrey) [reply CMakeLists.txt:196]: Eh only because GitHub doesn't have a setting for "can merge with review/approve others' PRs for merging, but not have full write access"

Comment (GMNGeoffrey) [reply CMakeLists.txt:196]: As a meta note. I think we should probably not be duplicating documentation between the CMake config and the markdown. Seems like we should be able to generate this (a task for a later time)

**Review by MaheshRavishankar — COMMENTED** [review]
I dont fully understand all the CMake here, but +1 on IREE_DEV_MODE. Not approving for someone with more know-how to actually approve.

File: docs/developers/get_started/cmake_options_and_variables.md [hunk 2]
```diff
@@ -54,7 +54,7 @@ Builds the IREE python bindings. Defaults to `OFF`.
 
 Builds the IREE TFLite C API compatibility shim. Defaults to `ON`.
 
-### `IREE_BUILD_BINDINGS_TFLITE_JAVA`:BOOL
+#### `IREE_BUILD_BINDINGS_TFLITE_JAVA`:BOOL
```
Comment (GMNGeoffrey) [inline docs/developers/get_started/cmake_options_and_variables.md:57]: Drive-by fix. This was at the wrong heading level

File: CMakeLists.txt (hunk 1 above)
Comment (marbre) [reply CMakeLists.txt:196]: > Eh only because GitHub doesn't have a setting for "can merge with review/approve others' PRs for merging, but not have full write access"

No problem at all. It's fine the way it is :)

> As a meta note. I think we should probably not be duplicating documentation between the CMake config and the markdown. Seems like we should be able to generate this (a task for a later time)

I think the docs are more expressive than the strings in the CMake config. But if we can, we should do.

**Review by ScottTodd — APPROVED** [review]
